### DISCUSSION TITLE
Gpu Backed Canvas, Do i need to do any configuration for CrossPlatform?

### DISCUSSION THREAD

**Discussion by cangunaydin** [AvaloniaUI/Avalonia#14245] [General]
Hello,
I am trying to use 2d canvas. I have based my app according to customskiapage in the samples.
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/CustomSkiaPage.cs

the demo works fine. I created xplatform app from scratch and apply the code. 
I have deployed the app to android. i have an old samsung galaxy a tablet (Galaxy Tab A (2016)). When i use developer options to show gpu usage it doesn't seem it uses gpu at all. And it seems it has little bit lag. So the question is how can i understand if drawing canvas is backed by gpu or not?
And also do i need to do some configuration to support gpu rendering?
thanks.

**Comment by maxkatz6** [comment]
By default Android app will use EGL, so it should be hardware accelerated: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Android/Avalonia.Android/AndroidPlatform.cs#L58

**Reply by maxkatz6** [reply]
You can check Avalonia logs to see if EGL initialization fails for any reason
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.OpenGL/Egl/EglPlatformGraphics.cs#L35

**Reply by maxkatz6** [reply]
> And it seems it has little bit lag.

But in general, it's not really supersized for an old android tablet. These were pretty weak devices, like other 4xx snapdragons.

**Comment by cangunaydin** [comment]
Hello again,
I was playing around little bit. i have added LogToTrace method and on debug console at first start i am getting couple of these when i debug it.

![Screenshot 2024-01-18 at 16 53 38](https://github.com/AvaloniaUI/Avalonia/assets/3688369/893856f2-5189-41e1-af77-f37168337e84)

but afterwards no exception,does it start to render before window is initialized?
and another thing is when i debug i can see that on lease, grContext is set. Also i can see that it is opengl. But don't know if this is a sign that if it is hardware accelerated. 
![Screenshot 2024-01-18 at 16 38 36](https://github.com/AvaloniaUI/Avalonia/assets/3688369/27aeab25-9fa6-4745-aed0-e274b36e6d3b)


And one last thing, i am trying to create a simple media player, i have a method to decode frame and get it to bytearray. Afterwards i do couple of copying and resizing operations. I was thinking maybe that's why i see lots of cpu spike instead of gpu spike
I was wondering if you have any idea that if these operations are not good in terms of performance. here is what i do whenever i invalidate the surface.
```csharp
var leaseFeature = context.TryGetFeature<ISkiaSharpApiLeaseFeature>();
        if (leaseFeature == null)
            context.DrawGlyphRun(Brushes.Black, _noSkia);
        else
        {
            using var lease = leaseFeature.Lease();
            var canvas = lease.SkCanvas;
            canvas.Save();
            // make sure the canvas is blank
            canvas.Clear(SKColors.White);
            var canvasWidth = (int)Bounds.Width;
            var canvasHeight = (int)Bounds.Height;
            if (_type == DrawingType.Video)
            {
                using (var data = SKData.CreateCopy(_frameBuffer))
                using (var original = new SKBitmap())
                {
                    // create a new bitmap using the memory
                    original.InstallPixels(
                        new SKImageInfo(_videoFrameWidth.Value, _videoFrameHeight.Value, SKColorType.Rgba8888),
                        data.Data);
                    using (var resized = original.Resize(new SKImageInfo(canvasWidth, canvasHeight),
                               SKFilterQuality.Medium))
                    {
                        canvas.DrawBitmap(resized, 0, 0);
                    }
                }
            }
```

The problem with this code is on some videos it works fine, but on some even the invalidate method is triggered it doesn't draw, probably it can not catch up with the frequency (fps) so invalidate is called before it finishes the first invalidate call. Maybe this is all normal, since the device is very old but since the video that is not playing is only 30fps, probably i am doing sth wrong :) By the way the same code works fine on iphone 12. But i don't wanna trust on that since it can be more troublesome in the future. 

if you have any suggestions i would like to hear about it. Thanks for the help.

**Reply by kekekeks** [reply]
>and another thing is when i debug i can see that on lease, grContext is set. Also i can see that it is opengl. But don't know if this is a sign that if it is hardware accelerated.

If there is a GrContext, the canvas is hw-accelerated.


For content like video players, try using [`CompositionCustomVisual`](https://github.com/AvaloniaUI/Avalonia/blob/3911fe3cd1591d4a7436c400b61e46649cfde2a0/samples/ControlCatalog/Pages/CompositionPage.axaml.cs#L219-L285), it's better suited for your needs, i. e. in the linked example `OnAnimationFrameUpdate` is called on every frame on the render thread, so you can check if you have a new decoded video frame and decide if you want to redraw the visual. That way you can make your video decoder to run on a background thread too.

**Reply by cangunaydin** [reply]
Hello,
thanks for the example, what i wonder though is if the video has slower fps i need to wait for invalidate. So two options i am thinking of. 
1- I can delay the task OnAnimationFrameUpdate and invalidate(this seems weird cause background thread is gonna be so fast, i need to put them into buffer maybe, then i can get some memory problems, maybe queuing) 
2- I can decode according to fps and invalidate immediately. (if the process takes long probably it will decrease the fps)

so don't know how should i handle it. I am kind of rookie on this subject. so it is good learning path for me :)  
I will try and see how it goes with different approaches.

**Reply by kekekeks** [reply]
`OnAnimationFrameUpdate`  happens on the render thread which uses our render timer. There is no control over when it would happen. If you don't call `Invalidate` from that callback, no actual frame rendering would happen (unless there are other changes in the visual tree), so you can just skip calling it if you don't have a video frame available.
